File: src/test/resources/PreorderTraversal.java
import java.util.*;

import com.jalgoarena.type.TreeNode;

public class Solution {

    public int[] preorderTraversal(TreeNode root) {
        List<Integer> items = new ArrayList<>();
        preorder(root, items);
        return items.stream().mapToInt(i -> i).toArray();
    }

    private void preorder(TreeNode root, List<Integer> items) {
        if (root != null) {
            items.add(root.data);
            preorder(root.left, items);
            preorder(root.right, items);
        }
    }
}

File: src/test/resources/PreorderTraversal.kt
import com.jalgoarena.type.TreeNode
import java.util.*
import com.jalgoarena.type.*

class PreorderTraversal {

    fun preorderTraversal(root: TreeNode?): IntArray {
        val items = ArrayList<Int>()
        preorder(root, items)
        return items.toIntArray()
    }

    private fun preorder(root: TreeNode?, items: MutableList<Int>) {
        if (root != null) {
            items.add(root.data)
            preorder(root.left, items)
            preorder(root.right, items)
        }
    }
}
